### DISCUSSION TITLE
Connect over TLS

### DISCUSSION THREAD

**Discussion by shockogit** [genotrance/px#181] [Q&A]
Does PX proxy support clients connecting over TLS i.e. where a client may have setup `https_proxy=https://pxproxy:3128` for example as opposed to `http_proxy`?

**Comment by genotrance** [comment]
Px simply uses libcurl so anything is possible. That being said, when Px connects to an upstream NTLM/Kerberos proxy, it just connects using plain HTTP since folks usually setup http_proxy or equivalent.

Further, the actual SSL conversation for HTTPS URLs is directly between the client and the target server - Px does an HTTP CONNECT to the NTLM/Kerberos proxy which will further do an HTTP CONNECT to the target server and let the client directly talk to the target server. All the SSL key exchange and encryption/decryption happens between client and server. Px and upstream proxy have no idea what is being discussed and simply forward packets back and forth.

I'm not sure how TLS fits into all this but presume that also will be directly between the client and the target server. If you can test and find out, it will be best. Curious to know your results.

**Comment by shockogit** [comment]
I explained it poorly so I have updated my original post. My understanding is when an explicit proxy (PX in this case) is configured at a client by default it will send a HTTP connect to PX proxy e.g. `http_proxy=http://pxproxy:3128`. The CONNECT is sent in clear text over HTTP to the proxy. Some tools/clients support `https_proxy=https://pxproxy:3128` and this would be a connect over TLS if supported by the proxy.

**Reply by genotrance** [reply]
Today Px does not support HTTPS, mainly because Px is used to authenticate on your behalf so you should use it in a secure environment - either completely on your local machine or in a network that you trust so that others do not impersonate you.

Of course Px supports connecting from another host with `gateway` mode and can filter IPs that it talks to but doesn't yet support any authentication or secured incoming connections so there's room for improvement.

Adding HTTPS is not too hard but will then require the user to generate keys and maybe even trust the proxy or accept self-signed certs. More documentation headaches as well.

I'm curious what your use case is though and curious if this is useful enough to implement. Open to PRs as well of course.

**Reply by genotrance** [reply]
Px supports client authentication since v0.9.0. However, there's still no plan to enable HTTPS.
